keps/0002-kep-moveengine-hld.md

```diff
@@ -0,0 +1,152 @@
+---
+kep-number: 1
+title: MoveEngine - High Level Design
+authors:
+  - "@vitta"
+owners:
+  - Uma Mukkara
+  - Mayank
+editor: 
+creation-date: 2019-05-16
+last-updated: 2019-05-19
+status: provisional
+see-also:
+  - KEP-None
+replaces:
+  - KEP-None
+superseded-by:
+  - KEP-None
+---
+
+# KubeMove - High Level Design
+
+
+
+## Table of Contents
+
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Design acceptance criteria](#design-acceptance-criteria)
+* [Implementation History](#implementation-history)
+
+## Summary
+
+From the KubeMove HLD KEP, KubeMoveEngine CRD defines the placeholders for the workflow for a given application. The CR is created by the user by invoking the spec that has the following details
+   - Application to which this CR applies
+   - Target location including the namespace. User can specify multiple target locations, in which case multiple KMovePair CRs get created.
+   - The Dynamic Data Mobilizer or DDM details
+   - schedule time to transfer data at regular interval
+   - Telemetry management
+
+`MoveEngine` controller that is watching this CR will deploy the KubeMove framework(`MoveEngine`, `DataSync` controller) at remote cluster to
+setup the application and transfer the data to remote cluster. This controller creates the datasync CR periodically and invokes the `DataSync`
+controller to initiate the data transfer through DDM. Once DDM completes the data transfer, `MoveEngine` controller updates the remote `DataSync` controller by creating dataSync CR at remote cluster to validate/verify the data transfer, and to perform the restore of data.
+
+## Motivation
+
+This KEP is to get into details of MoveEngine CR
+
+### Goals
+
+This KEP is to get into details of MoveEngine CR
+
+### Non-Goals
+
+## Proposal
+
+`spec` of MoveEngine CR consists of following fields:
+```
+    # remoteCluster details
+    movePair: cluster-2
+
+    #namespaces which needs to be migrated, empty for default, all for all namespace
+    namespace: ns1
+
+    # targeted namespace, empty for source namespace
+    remoteNamespace: ns2
+
+    #label selectors
+    selectors:
```

Comment:
From our Stash experience, we found that it was better to use 1-1 mapping for MoveEngine CR and k8s workload.

In Stash v1alpha1, we use label selector based approach. It caused a number of problems:

- When do you resolve the labels? This causes a lot of complexity and confusing for the user.
- What if different types of apps have the same labels?
- When keeping track of `.status`, it is easier to have one workload to deal with instead of dynamic number of workloads.
- Also, users want to move one app at a time. So, separate CRs are easier.

So, in our Stash V1beta1, we switched to use `ref` instead labels selectors.https://github.com/stashed/stash/blob/master/apis/stash/v1beta1/types.go#L46-L88

Later you can introduce a `MoveEngineBlueprint` type non-namespaced CR so that if users have multiple workloads that need moving they can use the `MoveEngineBlueprint` as a template.